src/databricks/labs/ucx/assessment/sequencing.py

```diff
@@ -0,0 +1,261 @@
+# but we use list for the required mutability
+QueueEntry = list[int | MigrationNode | str]
+
+
+class PriorityQueue:
+    """A migration node priority queue.
+
+    Note:
+        This implementation does not support threading safety as that is not required.
+
+    Source:
+        See https://docs.python.org/3/library/heapq.html#priority-queue-implementation-notes on the changes below
+        to handle priority changes in the task. Also, the _UPDATED marker is introduced to avoid updating removed nodes.
+    """
+
+    _REMOVED = "<removed>"  # Mark removed items
+
+    def __init__(self):
+        self._entries: list[QueueEntry] = []
+        self._entry_finder: dict[MigrationNode, QueueEntry] = {}
+        self._counter = itertools.count()  # Tiebreaker with equal priorities, then "first in, first out"
+
+    def put(self, priority: int, task: MigrationNode) -> None:
+        """Put or update task in the queue.
+
+        The lowest priority is retrieved from the queue first.
+        """
+        if task in self._entry_finder:
+            self._remove(task)
+        count = next(self._counter)
+        entry = [priority, count, task]
+        self._entry_finder[task] = entry
+        heapq.heappush(self._entries, entry)
+
+    def get(self) -> MigrationNode | None:
+        """Gets the tasks with lowest priority."""
+        while self._entries:
+            _, _, task = heapq.heappop(self._entries)
+            if task == self._REMOVED:
+                continue
+            assert isinstance(task, MigrationNode)
+            self._remove(task)
+            # Ignore type because heappop returns Any, while we know it is an QueueEntry
+            return task
+        return None
+
+    def _remove(self, task: MigrationNode) -> None:
+        """Remove a task from the queue."""
+        entry = self._entry_finder.pop(task)
+        entry[2] = self._REMOVED
+
+
+class MigrationSequencer:
+    """Sequence the migration dependencies in order to execute the migration.
+
+    Similar to the other graph logic, we first build the graph by registering dependencies, then we analyse the graph.
+    Analysing the graph in this case means: computing the migration sequence in `meth:generate_steps`.
+    """
+
+    def __init__(self, ws: WorkspaceClient, admin_locator: AdministratorLocator):
+        self._ws = ws
+        self._admin_locator = admin_locator
+        self._last_node_id = 0
+        self._nodes: dict[MigrationNodeKey, MigrationNode] = {}
+        self._outgoing: dict[MigrationNodeKey, set[MigrationNode]] = defaultdict(set)
+
+    def register_workflow_task(self, task: jobs.Task, job: jobs.Job, _graph: DependencyGraph) -> MigrationNode:
+        task_id = f"{job.job_id}/{task.task_key}"
+        task_node = self._nodes.get(("TASK", task_id), None)
+        if task_node:
+            return task_node
+        job_node = self.register_workflow_job(job)
```

Comment:
@ericvergnaud : Why do we go from the task to the job instead of the other way around?